packages/lib/person/service.ts

```diff
@@ -229,65 +242,81 @@ export const updatePerson = async (personId: string, personInput: TPersonUpdateI
-          attributeClass: {
-            name: "userId",
-          },
-          value: userId,
-        },
-      },
-    },
-    select: selectPerson,
-  });
-
-  if (person) {
-    return transformPrismaPerson(person);
-  } else {
-    // Create a new person with the userId attribute
-    const userIdAttributeClass = await getAttributeClassByName(environmentId, "userId");
-
-    if (!userIdAttributeClass) {
-      throw new ResourceNotFoundError("Attribute class not found for the given environment", environmentId);
-    }
+export const getOrCreatePersonByUserId = async (userId: string, environmentId: string): Promise<TPerson> => {
+  const personPrisma = await unstable_cache(
+    async () => {
+      validateInputs([userId, ZString], [environmentId, ZId]);
 
-    const person = await prisma.person.create({
-      data: {
-        environment: {
-          connect: {
-            id: environmentId,
-          },
-        },
-        attributes: {
-          create: [
-            {
+      // Check if a person with the userId attribute exists
+      const personPrisma = await prisma.person.findFirst({
+        where: {
+          environmentId,
+          attributes: {
+            some: {
               attributeClass: {
-                connect: {
-                  id: userIdAttributeClass.id,
-                },
+                name: "userId",
               },
               value: userId,
             },
-          ],
+          },
         },
-      },
-      select: selectPerson,
-    });
+        select: selectPerson,
+      });
+
+      if (personPrisma) {
+        return personPrisma;
+      } else {
+        // Create a new person with the userId attribute
+        const userIdAttributeClass = await getAttributeClassByName(environmentId, "userId");
+
+        if (!userIdAttributeClass) {
+          throw new ResourceNotFoundError(
+            "Attribute class not found for the given environment",
+            environmentId
+          );
+        }
+
+        const person = await prisma.person.create({
+          data: {
+            environment: {
+              connect: {
+                id: environmentId,
+              },
+            },
+            attributes: {
+              create: [
+                {
+                  attributeClass: {
+                    connect: {
+                      id: userIdAttributeClass.id,
+                    },
+                  },
+                  value: userId,
+                },
+              ],
+            },
+          },
+          select: selectPerson,
+        });
 
-    if (person) {
-      // revalidate person
-      revalidateTag(person.id);
+        personCache.revalidate({
+          id: person.id,
+          environmentId: person.environmentId,
+          userId,
+        });
+
+        return person;
+      }
+    },
+    [`getOrCreatePersonByUserId-${userId}-${environmentId}`],
+    {
+      tags: [personCache.tag.byUserIdAndEnvironmentId(userId, environmentId)],
```

Comment:
I would here use these tags:
```
[personCache.tag.byUserId(userId), personCache.tag.byEnvironmentId(environmentId)],
```

We need to revalidate if either the person with this userId changed (because we need to return the modified object) or we need to revalidate if it was not found previously and now is in the environment to get it next time.
But with that argumentation we don't need to revalidate the cache from the outside since it's created running this method.
So the cache tag should only be
```
personCache.tag.byUserId(userId, environmentId)
```

Reply:
Updated to

```
[personCache.tag.byUserId(userId), personCache.tag.byEnvironmentId(environmentId)],
```